examples/raja-launch-mat-mat.cpp

```diff
@@ -0,0 +1,690 @@
+*/
+#if defined(RAJA_ENABLE_CUDA)
+__global__ void matMultKernel(int N, double* C, double* A, double* B)
+{
+  // Block row and column
+  const int by = blockIdx.y;
+  const int bx = blockIdx.x;
+
+   // Each thread computes one element of Csub
+  // by accumulating results into Cvalue
+  double Cvalue(0.0);
+
+  // Thread row and column within local Csub
+  const int ty = threadIdx.y; //local row
+  const int tx = threadIdx.x; //local column
+
+  const int row = by * CUDA_BLOCK_SIZE + ty;  // Matrix row index
+  const int col = bx * CUDA_BLOCK_SIZE + tx;  // Matrix column index
+
+  // Shared memory used to store Asub and Bsub respectively
+  __shared__ double As[CUDA_BLOCK_SIZE][CUDA_BLOCK_SIZE];
+  __shared__ double Bs[CUDA_BLOCK_SIZE][CUDA_BLOCK_SIZE];
+
+  // Loop over all the sub-matrices of A and B that are
+  // required to compute Csub
+  // Multiply each pair of sub-matrices together
+  // and accumulate the results
+  for (int m = 0; m < (N / CUDA_BLOCK_SIZE); ++m) {
+
+    // Load Asub and Bsub from device memory to shared memory
+    // Each thread loads one element of each sub-matrix
+    As[ty][tx] = A[row*N + m*CUDA_BLOCK_SIZE + tx];
+    Bs[ty][tx] = B[(m*CUDA_BLOCK_SIZE + ty)*N + col];
+
+    // Synchronize to make sure the sub-matrices are loaded
+    // before starting the computation
+    __syncthreads();
+
+    // Multiply Asub and Bsub together
+    for (int e = 0; e < CUDA_BLOCK_SIZE; ++e)
+      Cvalue += As[ty][e] * Bs[e][tx];
+
+    // Synchronize to make sure that the preceding
+    // computation is done before loading two new
+    // sub-matrices of A and B in the next iteration
+    __syncthreads();
+  }
+
+  // Write Csub to device memory
+  // Each thread writes one element
+  C[col + N*row] = Cvalue;
+
+}
+#endif
+
+//
+// Functions for checking results
+//
+template <typename T>
+void checkResult(T *C, int N);
+
+template <typename T>
+void checkResult(RAJA::View<T, RAJA::Layout<DIM>> Cview, int N);
+
+//
+// Functions for printing results
+//
+template <typename T>
+void printResult(T *C, int N);
+
+template <typename T>
+void printResult(RAJA::View<T, RAJA::Layout<DIM>> Cview, int N);
+
+
+namespace RAJA {
+
+  enum ExecPlace {
+    HOST,
+    DEVICE
+  };
+
+  template<ExecPlace EXEC_PLACE, typename POLICY>
+  struct LPolicy{
+    static constexpr ExecPlace exec_place = EXEC_PLACE;
+    using policy_t = POLICY;
+  };
+
+  struct Teams{
+    int value[3];
+
+    Teams() : value{1,1,1}{}
+
+    Teams(int i) : value{i,1,1}{}
+
+    Teams(int i, int j) : value{i,j,1}{}
+
+    Teams(int i, int j, int k) : value{i,j,k}{}
+  };
+
+  struct Threads{
+    int value[3];
+
+    Threads() : value{1,1,1}{}
+
+    Threads(int i) : value{i,1,1}{}
+
+    Threads(int i, int j) : value{i,j,1}{}
+
+    Threads(int i, int j, int k) : value{i,j,k}{}
+  };
+
+  struct Lanes{
+    int value;
+
+    Lanes() : value(0){}
+
+    Lanes(int i) : value(i){}
+  };
+
+  class ResourceBase {
```

Comment:
Does this define a hierarchy of parallelization constructs, coarsest grained to finest grained?

Reply:
yes

Reply:
Well, OK then.